File: src/C3.Infrastructure/Preferences/LegacySettingsProfileReadResult.vb
Namespace Preferences

    Public Enum LegacySettingsProfileReadFailure
        None = 0
        Unavailable
        TooLarge
        MalformedXml
        InvalidStructure
        DuplicateSetting
        InvalidValue
    End Enum

    Public NotInheritable Class LegacySettingsProfileReadResult

        Private Sub New(
                candidate As LegacySettingsProfileCandidate,
                profile As LegacyUserSettingsProfile,
                failure As LegacySettingsProfileReadFailure,
                failureMessage As String,
                failureException As Exception)

            Me.Candidate = candidate
            Me.Profile = profile
            Me.Failure = failure
            Me.FailureMessage = failureMessage
            Me.FailureException = failureException
        End Sub

        Public ReadOnly Property Candidate As LegacySettingsProfileCandidate

        Public ReadOnly Property Profile As LegacyUserSettingsProfile

        Public ReadOnly Property Failure As LegacySettingsProfileReadFailure

        Public ReadOnly Property FailureMessage As String

        Public ReadOnly Property FailureException As Exception

        Public ReadOnly Property IsSuccess As Boolean
            Get
                Return Profile IsNot Nothing
            End Get
        End Property

        Friend Shared Function Succeeded(
                profile As LegacyUserSettingsProfile) As LegacySettingsProfileReadResult

            Return New LegacySettingsProfileReadResult(
                profile.Candidate,
                profile,
                LegacySettingsProfileReadFailure.None,
                Nothing,
                Nothing)
        End Function

        Friend Shared Function Failed(
                candidate As LegacySettingsProfileCandidate,
                failure As LegacySettingsProfileReadFailure,
                message As String,
                Optional failureException As Exception = Nothing) As LegacySettingsProfileReadResult

            Return New LegacySettingsProfileReadResult(
                candidate,
                Nothing,
                failure,
                message,
                failureException)
        End Function

    End Class

End Namespace

File: src/C3.Infrastructure/Preferences/LegacySettingsProfileReadResult.cs
using System;
using System.Runtime.CompilerServices;

namespace C3.Infrastructure.Preferences
{
	public sealed class LegacySettingsProfileReadResult
	{
		public LegacySettingsProfileCandidate Candidate
		{
			get;
		}

		public LegacyUserSettingsProfile Profile
		{
			get;
		}

		public LegacySettingsProfileReadFailure Failure
		{
			get;
		}

		public string FailureMessage
		{
			get;
		}

		public Exception FailureException
		{
			get;
		}

		public bool IsSuccess => Profile != null;

		private LegacySettingsProfileReadResult(LegacySettingsProfileCandidate candidate, LegacyUserSettingsProfile profile, LegacySettingsProfileReadFailure failure, string failureMessage, Exception failureException)
		{
			Candidate = candidate;
			Profile = profile;
			Failure = failure;
			FailureMessage = failureMessage;
			FailureException = failureException;
		}

		internal static LegacySettingsProfileReadResult Succeeded(LegacyUserSettingsProfile profile)
		{
			return new LegacySettingsProfileReadResult(profile.Candidate, profile, LegacySettingsProfileReadFailure.None, null, null);
		}

		internal static LegacySettingsProfileReadResult Failed(LegacySettingsProfileCandidate candidate, LegacySettingsProfileReadFailure failure, string message, Exception failureException = null)
		{
			return new LegacySettingsProfileReadResult(candidate, null, failure, message, failureException);
		}
	}
}
